src/Jellyfin.Plugin.ListenBrainz/Clients/ListenBrainzClient.cs

```diff
@@ -233,12 +233,12 @@ private IEnumerable<Listen> ToListens(IEnumerable<StoredListen> storedListens)
         if (_libraryManager is null) throw new InvalidOperationException("Library manager is not available");
 
         var listensToConvert = storedListens.ToArray();
-        return listensToConvert
-            .Select(l => _libraryManager.GetItemById(l.Id))
-            .Cast<Audio>()
-            .Select(a => a.AsListen(
-                listensToConvert.First(l => l.Id == a.Id).ListenedAt,
-                listensToConvert.First(l => l.Id == a.Id).Metadata));
+        return listensToConvert.Select(l =>
+        {
+            var baseItem = _libraryManager.GetItemById(l.Id);
+            var audio = (Audio?)baseItem ?? throw new PluginException("Invalid item");
```

Comment:
I'm not a fan of throwing an exception here, but the previous behavior with the `Cast<Audio>()` was quite similar. Maybe it's possible to skip missing tracks and keep them in the cache, or move them to some “error” file. But I think this refactor should be separate from this PR.

Reply:
Yeah, let's keep it as is for now.

I know the code is pretty much garbage, even after I rewrote everything for v3.x.. 😅 The conversion to Listens should not be done in the `SendListensAsync` anyway, and after it's thrown out, it would allow granular filtering and then skipping over listens of invalid items. But that's definitely for another time, I agree.